### PR TITLE
Authenticate customer account extension requests

### PR DESCRIPTION
**Pull request by lizkenyon**
### WHY are these changes introduced?

Introduces a new API to authenticate requests from[ Customer Account Extensions](https://shopify.dev/docs/apps/customer-accounts).

* Customer account extensions can use networks access to make HTTP requests to apps, and other external sources. 
* These requests can use [session tokens to be authenticated](https://shopify.dev/docs/api/customer-account-ui-extensions/unstable/apis/session-token#examples) 

Because Checkout UI extensions and Customer Account extensions are authenticated the same way, this refactors the the authenticateCheckoutFactory to be useable for both.

TODO
* ~generate docs~
* ~generate changelog~
* ~test in dev app~

## Type of change

- [ ] Patch: Bug (non-breaking change which fixes an issue)
- [x] Minor: New feature (non-breaking change which adds functionality)
- [ ] Major: Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have used `yarn changeset` to create a draft changelog entry (do NOT update the `CHANGELOG.md` files manually)
- [x] I have added/updated tests for this change
- [x] I have documented new APIs/updated the documentation for modified APIs (for public APIs)

### CODE REVIEW

File: packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/authenticate.public.customerAccount.doc.example.offers.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// Most apps would load this from their database
+export function getOffers(shop: string, customerID: string) {
+    const offers: Record<string, any[]> = {
```
Comment (paulomarg) [inline packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/authenticate.public.customerAccount.doc.example.offers.ts:3]: Nit, but is the tab width right here?

File: packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/authenticate.public.customerAccount.doc.example.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import type {ActionFunctionArgs, LoaderFunctionArgs} from '@remix-run/node';
+import {json} from '@remix-run/node';
+
+import {authenticate} from '../shopify.server';
+import {getOffers} from '../offers.server';
+
+// The loader responds to preflight requests from Shopify
+export const loader = async ({request}: LoaderFunctionArgs) => {
+  await authenticate.public.checkout(request);
```
Comment (paulomarg) [inline packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/authenticate.public.customerAccount.doc.example.ts:9]: Wrong function name :)

File: packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/types.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+   * <caption>Using the decoded session token.</caption>
+   * <description>Get store-specific data using the `sessionToken` object.</description>
+   * ```ts
+   * // app/routes/public/my-route.ts
+   * import { LoaderFunctionArgs, json } from "@remix-run/node";
+   * import { authenticate } from "../shopify.server";
+   * import { getMyAppData } from "~/db/model.server";
+   *
+   * export const loader = async ({ request }: LoaderFunctionArgs) => {
+   *   const { sessionToken } = await authenticate.public.checkout(
```
Comment (paulomarg) [inline packages/apps/shopify-app-remix/src/server/authenticate/public/customer-account/types.ts:33]: Wrong method here too.

File: packages/apps/shopify-app-remix/src/server/authenticate/public/factory.ts [hunk 4]
```diff
@@ -45,5 +49,5 @@ export function authenticatePublicFactory<
   authenticatePublic.checkout = authenticateCheckout;
   authenticatePublic.appProxy = authenticateAppProxy;
 
-  return authenticatePublic;
+  return authenticatePublic as AuthenticatePublic<Future>;
```
Comment (paulomarg) [inline packages/apps/shopify-app-remix/src/server/authenticate/public/factory.ts:52]: Why do we need to cast the type here? I would think that we're returning the appropriate type given the definitions.

File: packages/apps/shopify-app-remix/src/server/authenticate/public/extension/authenticate.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  respondToBotRequest,
+  respondToOptionsRequest,
+  getSessionTokenHeader,
+  validateSessionToken,
+  ensureCORSHeadersFactory,
+} from '../../helpers';
+
+import {AuthenticateExtension, ExtensionContext} from './types';
+
+export function authenticateExtensionFactory(
```
Comment (paulomarg) [inline packages/apps/shopify-app-remix/src/server/authenticate/public/extension/authenticate.ts:13]: I wonder if we can touch up the types a little bit so:
- we don't have this little bit of duplication
- we don't have to cast the types

Maybe we can have something like:

```suggestion
export function authenticateExtensionFactory<Options, Context>(): (Request, Options) => Promise<Context> { ... }
```

that way we wouldn't need the generic empty types, and the types returned by this function could be used without casting.

File: packages/apps/shopify-app-remix/src/server/authenticate/public/factory.ts (hunk 4 above)
Comment (lizkenyon) [reply packages/apps/shopify-app-remix/src/server/authenticate/public/factory.ts:52]: We do also cast  the return above,
 I don't know if typescript was just struggling with evaluating the FF type in this scenario, but I was getting an error.

**Pull request merged**